File: Class/Practise 4.1/src/PizzaBuilder.cpp
#include "PizzaBuilder.h"
#include <iostream>
#include <print>

PizzaBuilder::PizzaBuilder(const IngredientsReader& reader, const std::string& filename) {
    availableIngredients = reader.readIngredients(filename);
    this->reset();
}

void PizzaBuilder::showIngredients() const {
    std::println("Available ingredients:");
    for (const auto& pair : availableIngredients) {
        std::println("- {} (${})", pair.first, pair.second.getCost());
    }
}

void PizzaBuilder::addIngredientToPizza(const std::string& ingredientName) {
    if (availableIngredients.find(ingredientName) != availableIngredients.end()) {
        pizza->addIngredient(availableIngredients.at(ingredientName));
    } else {
        std::println("Ingredient not found: {}", ingredientName);
    }
}

std::unique_ptr<Pizza> PizzaBuilder::getProduct() {
    auto result = std::move(pizza);
    this->reset();
    return result;
}

void PizzaBuilder::reset() {
    this->pizza = std::make_unique<Pizza>();
}

std::unordered_map<std::string, Ingredient> PizzaBuilder::getAvailableIngredients() const {
    return availableIngredients;
}

File: Home/Home 4.1/PizzaBuilder.swift
import Foundation

class PizzaBuilder {
    private var pizza: Pizza
    private var availableIngredients: [String: Ingredient]

    init(reader: IngredientsReader, filename: String) {
        self.pizza = Pizza()
        self.availableIngredients = reader.readIngredients(from: filename)
    }

    func showIngredients() {
        print("Available ingredients:")
        for (name, ingredient) in availableIngredients {
            print("- \(name) ($\(ingredient.cost))")
        }
    }

    func addIngredientToPizza(_ ingredientName: String) {
        if let ingredient = availableIngredients[ingredientName] {
            pizza.addIngredient(ingredient)
        } else {
            print("Ingredient not found: \(ingredientName)")
        }
    }

    func getProduct() -> Pizza {
        let result = pizza
        self.pizza = Pizza()
        return result
    }
}
